crates/iota-metrics/src/hardware_metrics.rs

```diff
@@ -0,0 +1,283 @@
+    let mut unit_idx = 0;
+    // Shift right until we're below 1024^2 or reach end of units
+    while value >= (1024 * 1024) && unit_idx < UNITS.len() - 2 {
+        value >>= 10;
+        unit_idx += 1;
+    }
+    let value: f64 = value as f64 / 1024.0;
+    unit_idx += 1;
+
+    format!("{:.2} {}", value, UNITS[unit_idx])
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum HardwareMetricsErr {
+    #[error("Failed creating metric")]
+    ErrCreateMetric(prometheus::Error),
+    #[error("Failed registering hardware metrics onto RegistryService")]
+    ErrRegisterHardwareMetrics(prometheus::Error),
+}
+
+#[cfg(test)]
+mod tests {
+    use std::{
+        net::SocketAddrV4,
+        time::{SystemTime, UNIX_EPOCH},
+    };
+
+    use super::*;
+
+    #[test]
+    fn test_hardware_specs() {
+        let hardware_specs = HardwareSpecs::new().unwrap();
+        let metric_families = hardware_specs.collect();
+        assert_eq!(&metric_families.len(), &4);
+    }
+
+    #[tokio::test]
+    async fn test_collect_hardware_specs() -> Result<(), String> {
+        let prom_server_addr: SocketAddrV4 = "0.0.0.0:9194".parse().unwrap();
+
+        let mut registry_svc = crate::start_prometheus_server(prom_server_addr.into());
+
+        register_hardware_metrics(&mut registry_svc).expect("Failed registering hardware metrics");
+
+        let now = SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .unwrap()
+            .as_millis() as i64;
+
+        let mut metric_families = registry_svc.gather_all();
+        for mf in metric_families.iter_mut() {
+            for m in mf.mut_metric() {
+                m.set_timestamp_ms(now);
+            }
+        }
+
+        let find_metric_label = |family_name: &str, label_name: &str| -> String {
+            let metric = metric_families
+                .iter()
+                .find(|mf| mf.get_name() == family_name)
+                .unwrap()
+                .get_metric()
+                .first()
+                .unwrap();
+            metric
+                .get_label()
+                .iter()
+                .find(|l| l.get_name() == label_name)
+                .unwrap()
+                .get_value()
+                .to_string()
+        };
+
+        assert_eq!(&metric_families.len(), &4);
+
+        let core_count = find_metric_label("cpu_specs", "cpu_core_count")
+            .parse::<usize>()
+            .unwrap();
+        assert!(core_count > 0 && core_count < 513);
+
+        let mem_total_human = find_metric_label("memory_specs", "memory_total_human");
+        assert!(
+            mem_total_human.ends_with(" GB"),
+            "memory is usually in the GB range at the moment"
+        );
+
+        let disk_total_human = find_metric_label("disk_specs", "disk_total_space_human");
+        assert!(
+            disk_total_human.ends_with(" GB") || disk_total_human.ends_with(" TB"),
+            "disk size is usually in GB or TB at the moment"
+        );
```

Comment:
I don't know about these tests, they seem like they could start failing randomly. Maybe that's very unlikely but I'd still rather have something future-proof